### PR TITLE
cast maps to lists, for params coming from a Phoenix form

### PR DESCRIPTION
**Pull request by egze**
Able to cast params coming from a nested Phoenix form.

Assuming we have the following schema for a quiz with a list of questions. A question is a combination of the actual question and a correct answer.

```json
{
  "title": "Quiz",
  "type": "object",
  "properties": {
    "questions": {
      "type": "array",
      "items": {
        "type": "object",
        "required": [
          "question",
          "answer"
        ],
        "properties": {
          "question": {
            "type": "string"
          },
          "answer": {
            "type": "string"
          }
        }
      }
    }
  }
}
```

To represent it in a Phoenix form, the HTML would be similar to this:

```html
<form>
  <div class="question">
    <input type="text" name="quiz[questions][0][question]" value="1 + 1 = ?" />
    <input type="text" name="quiz[questions][0][answer]" value="2" />
  </div>
  <div class="question">
    <input type="text" name="quiz[questions][1][question]" value="2 + 2 = ?" />
    <input type="text" name="quiz[questions][1][answer]" value="4" />
  </div>
</form>
```

The template for adding new questions dynamically to the form would be:
```html
<div class="question">
  <input type="text" name="quiz[questions][__INDEX__][question]"  />
  <input type="text" name="quiz[questions][__INDEX__][answer]" />
</div>
```
We would append the fragment to the form and replace `__INDEX__` dynamically with JS.

OK, now we submit the form, the params would be in this shape:

```elixir
params = %{
  "quiz" => %{
    "questions" => %{
      "0" => %{"question" => "1 + 1 = ?", "answer" => "2"},
      "1" => %{"question" => "2 + 2 = ?", "answer" => "4"}
    }
  }
}
```

With the change in the PR, after `Xema.cast(xema, params)` for the example above, the result would be:

```elixir
iex> Xema.cast(xema, params["quiz"])

{:ok, %{
  "questions" => [
    %{"answer" => "2", "question" => "1 + 1 = ?"},
    %{"answer" => "4", "question" => "2 + 2 = ?"}
  ]
}}
```

### CODE REVIEW

**Comment by egze** [comment]
For some reason, it doesn't cast the inner-map values. See the failing test: [test/xema/cast/list_test.exs:371](https://github.com/hrzndhrn/xema/pull/167/files#diff-1301f9cc662344773c19d0681d90c09bcc218293d3256fbf7bba66180bf18c25R371)

@NickNeck Do you have an idea why?

**Comment by NickNeck** [comment]
I will take a look next week.

**Comment by egze** [comment]
Hi @NickNeck Have you had a chance to look at the issue yet? Much appreciated 🙏

**Comment by NickNeck** [comment]
Sorry for my late response and thanks for your PR.
Most of the code looks good.

I would rather limit the cast here to maps with integer and integer-strings. That satisfies the case for maps from Phoenix forms. I do not think it makes sense to try casting any possible map.

The failing test happens because the values of the map are not correct cast. I have forgotten to tell you about the private functions `Xema.cast_values`. This functions are casting nested data in `tuple`, `keyword`, `list`, `struct` and `map`. Here we need a function like `defp cast_values(%Schema{items: items}, data, opts, path) when is_map(data) do` that works similar like `defp cast_values_list(%Schema{items: items} = schema, data, opts, path) when is_list(data) do`.

The missing function (so far): 
```elixir
  defp cast_values(%Schema{items: items} = schema, data, opts, path)
       when not is_nil(items) and is_map(data) do
    case items do
      nil ->
        {:ok, data}

      %Schema{} = schema ->
        data
        |> Enum.reduce_while([], fn {index, item}, acc ->
          case do_cast(schema, item, opts, [index | path]) do
            {:ok, cast} -> {:cont, [{index, cast} | acc]}
            {:error, _} = error -> {:halt, error}
          end
        end)
        |> case do
          {:error, _} = error -> error
          values -> {:ok, Map.new(values)}
        end

      items ->
        raise "coming soon"
    end
  end
```

**Comment by egze** [comment]
Thanks for the function. I removed the following part from it:
```elixir
nil -> {:ok, data}
```
because it is already handled by a guard.

I have 2 questions:

1) When you say
> I would rather limit the cast here to maps with integer and integer-strings.

do you mean by that, that the function should raise unless the key can be parsed to an Integer?

2) What case did you have in mind with this default match?
```elixir
items ->
  raise "coming soon"
```
What schema and data do I need to have to reach this code?

**Comment by NickNeck** [comment]
Moin

> do you mean by that, that the function should raise unless the key can be parsed to an Integer?

In castable the return value should be `{:ok, new_value}` or `{:error, %{to: :map, value: old_value}}`. This values will be handled in `Xema.cast`.

> 2. What case did you have in mind with this default match?

This case deals with validations where a schema has been specified for each item in the list (see https://github.com/hrzndhrn/xema/blob/main/test/xema/cast/list_test.exs#L266).

 I have updated `defp cast_values(%Schema{type: :list, items: items} = schema, ...` and add `Xema.Utils.to_sorted_list(map)` and added some more tests.

Have a nice Sunday.

**Comment by egze** [comment]
@NickNeck Moin moin. Thanks for the explanations. I fixed the last Credo warning, which was preventing the CI from being green.

Is there anything left from your end? I'm personally very happy how it turned out.

File: lib/xema/castable.ex [hunk 1]
```diff
@@ -302,8 +302,40 @@ defimpl Xema.Castable, for: Map do
     end)
   end
 
+  def cast(map, :list, _module, _schema) do
+    sorted_keys = map |> Map.keys() |> sort_keys()
+
+    {:ok, Enum.map(sorted_keys, fn k -> Map.get(map, k) end)}
```
Comment (NickNeck) [inline lib/xema/castable.ex:308]: One last point:
```suggestion
    with :error <- to_sorted_list(map, keys: false) do
      {:error, %{to: :map, value: map}}
    end
```
That needs also 
```elixir
defimpl Xema.Castable, for: Map do
  use Xema.Castable.Helper

  import Xema.Utils, only: [to_sorted_list: 2]
  ...
```
some lines above.

And, then the test `cast/2 with list of maps (Xema) sorts the list based on key mixed values` needs to be updated.

And then everything is fine.

Comment (egze) [reply lib/xema/castable.ex]: Thanks. Done in de75041d34791d84bf0f676ce07f57821bdcb1c1

**Comment by NickNeck** [comment]
Thank you very much for your contribution.
I will make a new release the next days.
❤️

**Pull request merged**